extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java

```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
+            TenantConfigContext resolvedContext, AccessToken codeAccessToken) {
+
+        if (codeAccessToken != null) {
+            vertxContext.put("code_flow_access_token_result", codeAccessToken);
         }
 
-        final JsonObject userInfo = resolvedContext.oidcConfig.authentication.isUserInfoRequired()
-                ? getUserInfo(vertxContext, request, resolvedContext)
-                : null;
+        Uni<JsonObject> userInfo = getUserInfoUni(vertxContext, request, resolvedContext);
 
-        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
-            @Override
-            public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+        return userInfo.onItem().transformToUni(
+                new Function<JsonObject, Uni<? extends SecurityIdentity>>() {
+                    @Override
+                    public Uni<SecurityIdentity> apply(JsonObject userInfo) {
+                        return createSecurityIdentityWithOidcServerUni(vertxContext, request, resolvedContext, userInfo);
+                    }
+                });
+    }
+
+    private Uni<SecurityIdentity> createSecurityIdentityWithOidcServerUni(RoutingContext vertxContext,
+            TokenAuthenticationRequest request, TenantConfigContext resolvedContext, final JsonObject userInfo) {
 
-                resolvedContext.auth.decodeToken(request.getToken().getToken(),
-                        new Handler<AsyncResult<AccessToken>>() {
+        final TokenCredential tokenCred = request.getToken();
+        if (tokenCred instanceof AccessTokenCredential && ((AccessTokenCredential) tokenCred).isOpaque()) {
+            // remote introspection is required, a blocking call
+            if (BlockingOperationControl.isBlockingAllowed()) {
+                return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                    @Override
+                    public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                        createSecurityIdentityWithOidcServer(uniEmitter, vertxContext, request, resolvedContext, userInfo);
+                    }
+                });
+            } else {
+                return Uni.createFrom().emitter(
+                        new Consumer<UniEmitter<? super SecurityIdentity>>() {
                             @Override
-                            public void handle(AsyncResult<AccessToken> event) {
-                                if (event.failed()) {
-                                    uniEmitter.fail(new AuthenticationFailedException(event.cause()));
-                                    return;
-                                }
+                            public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                                tenantResolver.getBlockingExecutor().execute(new Runnable() {
```

Comment:
Any reason to not use Infrastructure.getDEfaultWorkerPool()?

Reply:
@cescoffier It is the same executor which Stuart uses [here](https://github.com/quarkusio/quarkus/blob/master/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusIdentityProviderManagerImpl.java#L54) so I'm just trying to keep it consistent. Perhaps you'd like to open an issue for Stuart (or I can do it) to replace this executor with `Infrastructure.getDEfaultWorkerPool()` ?

Reply:
it's actually the same pool.